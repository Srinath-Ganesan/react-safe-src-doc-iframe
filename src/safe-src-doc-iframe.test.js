import React from 'react';
import { shallow, mount } from 'enzyme';
import SafeSrcDocIframe from './safe-src-doc-iframe';

describe('<SafeSrcDocIframe />', () => {
  const mockDocument = `
  <!DOCTYPE html>
  <html>
    <head></head>
    <body>
      mock html page
    </body>
  </html>
  `;

  const testIframe = () => <SafeSrcDocIframe title='some-title' srcDoc={ mockDocument } />;

  const renderShallowSafeframe = () => {
    return shallow(testIframe());
  };

  const renderMountedSafeIframe = () => {
    return mount(testIframe());
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });
  it('renders as expected', () => {
    expect(
      renderShallowSafeframe()
    ).toMatchSnapshot();
  });
  it('will not configure onload for iframe during componentDidMount if ref does not exist', () => {
    const wrapper = renderMountedSafeIframe();
    expect(() =>
      wrapper.instance().componentDidMount.call({})
    ).not.toThrow();
  });
  it('will configure onload for iframe', () => {
    const applySafeguardsSpy = jest.spyOn(
      SafeSrcDocIframe.prototype,
      'applySafeguards'
    );
    const wrapper = renderMountedSafeIframe();
    expect(wrapper.instance().iframeElement.onload).toBeDefined();
    wrapper.instance().iframeElement.onload();
    expect(applySafeguardsSpy).toHaveBeenCalled();
  });
  it('will not apply safeguard styles if contentDocument does not exist', () => {
    const wrapper = renderMountedSafeIframe();
    const disableStyleTag = document.createElement('style');
    expect(() =>
      wrapper.instance().applySafeguards.call({
        iframeElement: {},
        disableStyleTag
      })
    ).not.toThrow();
  });
  it('will not apply safeguard styles if contentDocument.body does not exist', () => {
    const wrapper = renderMountedSafeIframe();
    const disableStyleTag = document.createElement('style');
    const mockContentDocument = {
      getElementsByTagName: jest.fn().mockReturnValue([])
    };
    expect(() =>
      wrapper.instance().applySafeguards.call({
        iframeElement: {
          contentDocument: mockContentDocument
        },
        disableStyleTag
      })
    ).not.toThrow();
    expect(mockContentDocument.getElementsByTagName).toHaveBeenCalledWith(
      'body'
    );
  });
  it('will apply safeguard styles if contentDocument and contentDocument.body exists', () => {
    const wrapper = renderMountedSafeIframe();
    const mockBody = {
      appendChild: jest.fn()
    };
    const mockContentDocument = {
      getElementsByTagName: jest.fn().mockReturnValue([mockBody])
    };
    const disableStyleTag = document.createElement('style');
    wrapper.instance().applySafeguards.call({
      iframeElement: {
        contentDocument: mockContentDocument
      },
      disableStyleTag
    });
    expect(mockBody.appendChild).toHaveBeenCalledWith(disableStyleTag);
  });
  it('will dereference the style tag and text node when the component will unmount', () => {
    const wrapper = renderMountedSafeIframe();
    expect(wrapper.instance().disableStyleTag).toBeDefined();
    expect(wrapper.instance().disableStylesTextNode).toBeDefined();
    wrapper.instance().componentWillUnmount();
    expect(wrapper.instance().disableStyleTag).toBe(null);
    expect(wrapper.instance().disableStylesTextNode).toBe(null);
    wrapper.unmount();
  });
});
