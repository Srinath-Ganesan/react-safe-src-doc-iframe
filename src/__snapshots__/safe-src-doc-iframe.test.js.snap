// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SafeSrcDocIframe /> renders as expected 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SafesrcDocIframe
    referrerPolicy="no-referrer"
    sandbox="allow-same-origin"
    src=""
    srcDoc="
  <!DOCTYPE html>
  <html>
    <head></head>
    <body>
      mock html page
    </body>
  </html>
  "
    title="some-title"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "referrerPolicy": "no-referrer",
      "sandbox": "allow-same-origin",
      "srcDoc": "
  <!DOCTYPE html>
  <html>
    <head></head>
    <body>
      mock html page
    </body>
  </html>
  ",
      "title": "some-title",
    },
    "ref": [Function],
    "rendered": null,
    "type": "iframe",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "referrerPolicy": "no-referrer",
        "sandbox": "allow-same-origin",
        "srcDoc": "
  <!DOCTYPE html>
  <html>
    <head></head>
    <body>
      mock html page
    </body>
  </html>
  ",
        "title": "some-title",
      },
      "ref": [Function],
      "rendered": null,
      "type": "iframe",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
